### PR TITLE
[WIP]Copy inputs and labels for RecurrentNetworkTest failure on Windows VS #2221

### PR DESCRIPTION
**Pull request by kartikdutt18**
The error still persists but I think this will be easier for people to suggest changes.
Current changes:
1. Create a copy for input data, labels in RNN tests.
2. Create a copy in RNN layer for input(every input i.e. all data steps even though they will occupy same size)

### CODE REVIEW

File: src/mlpack/methods/ann/layer/lstm_impl.hpp [hunk 1]
```diff
@@ -167,7 +167,10 @@ void LSTM<InputDataType, OutputDataType>::Forward(
 {
   //! Locally-stored cellState.
   OutputType cellState;
-  Forward(std::move(input), std::move(output), std::move(cellState), false);
+  // Trying a copy here as done in other layers rather than using std::move.
+  inputTemp = arma::mat(input.memptr(),
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer/lstm_impl.hpp]: I don't think this is required.

**Comment by kartikdutt18** [comment]
Hi @zoq, Could you give a quick review for this PR.

File: src/mlpack/methods/ann/layer/lstm_impl.hpp (hunk 1 above)
Comment (zoq) [reply src/mlpack/methods/ann/layer/lstm_impl.hpp]: You mean `inputTemp`?

File: src/mlpack/methods/ann/layer/lstm_impl.hpp [hunk 2]
```diff
@@ -276,7 +279,7 @@ void LSTM<InputDataType, OutputDataType>::Forward(InputType&& input,
       (forwardStep + batchSize) * outSize, outSize, batchSize, false, false);
 
   cellState = OutputType(cell.memptr() +
-      forwardStep * outSize, outSize, batchSize, false, false);
+      forwardStep * outSize, outSize, batchSize, true);
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/lstm_impl.hpp]: Actually, there is no need to make a copy of internal parameter.

File: src/mlpack/tests/recurrent_network_test.cpp [hunk 3]
```diff
@@ -1464,8 +1464,8 @@ BOOST_AUTO_TEST_CASE(LargeRhoValueRnnTest)
     return result;
   };
 
-  std::vector<MatType> inputs(trainingData.size());
-  std::vector<MatType> targets(trainingData.size());
+  std::vector<MatType> inputs(trainingData.size(), MatType());
```
Comment (zoq) [inline src/mlpack/tests/recurrent_network_test.cpp]: Maybe this is the real problem?

**Comment by zoq** [comment]
Hm, looks like the build finished without any error, let's restart the job.

**Comment by zoq** [comment]
https://dev.azure.com/mlpack/mlpack/_build/results?buildId=727&view=results
https://dev.azure.com/mlpack/mlpack/_build/results?buildId=728&view=results

File: src/mlpack/methods/ann/layer/lstm_impl.hpp (hunk 1 above)
Comment (kartikdutt18) [reply src/mlpack/methods/ann/layer/lstm_impl.hpp]: Yes, I meant inputTemp, I think `std::move` should work fine.

File: src/mlpack/tests/recurrent_network_test.cpp (hunk 3 above)
Comment (kartikdutt18) [reply src/mlpack/tests/recurrent_network_test.cpp]: I am not really sure about this, Valgrind throws an error for `Uninitialized Value` error, this reduces the number of the the error but it the error for `Invalid read` as well as `Uninitialized Value` still persists.

**Comment by kartikdutt18** [comment]
Memory checks have failed but it still shows passed on [github](http://ci.mlpack.org/job/pull-requests%20mlpack%20memory/3627/valgrindResult/).

**Comment by kartikdutt18** [comment]
Hmm, There is no error log for valgrind this time(after my last commit).

File: src/mlpack/tests/recurrent_network_test.cpp [hunk 4]
```diff
@@ -817,15 +817,15 @@ void DistractedSequenceRecallTestNetwork(
...
-    arma::cube inputTemp, labelsTemp;
     for (size_t iteration = 0; iteration < (9 + offset); iteration++)
     {
       for (size_t j = 0; j < trainDistractedSequenceCount; j++)
       {
+        arma::cube inputTemp, labelsTemp;
         inputTemp = arma::cube(trainInput.at(0, j).memptr(), inputSize, 1,
-            trainInput.at(0, j).n_elem / inputSize, false, true);
+            trainInput.at(0, j).n_elem / inputSize, true, true);
         labelsTemp = arma::cube(trainLabels.at(0, j).memptr(), outputSize, 1,
```
Comment (kartikdutt18) [inline src/mlpack/tests/recurrent_network_test.cpp]: So I ran valgrind plus gdb on this PR, one of the backtrace leads me to believe that the invalid read occurs here.

Comment (rcurtin) [reply src/mlpack/tests/recurrent_network_test.cpp]: Interesting, does the change you made to copy the auxiliary memory seem to help?  Or does it still occur despite that?

Comment (kartikdutt18) [reply src/mlpack/tests/recurrent_network_test.cpp]: It occurs even after creating a copy, should I run it (Valgrind + gdb) without creating a copy to see if creating a copy does help?

Comment (kartikdutt18) [reply src/mlpack/tests/recurrent_network_test.cpp]: I think I will do a rebase and then run valgrind + gdb again. I think the recent merge plus creating a copy might help, not sure though.

Comment (rcurtin) [reply src/mlpack/tests/recurrent_network_test.cpp]: :+1: so, one of the things that could be happening here that could make valgrind point out this line in specific is if some memory gets allocated to make `inputTemp` here, but then that memory is, e.g., passed somewhere else but `inputTemp` goes out of scope and the memory gets freed.

I'm not sure if that's happening here though.  I hope that makes sense, here's a MWE:

```
int main()
{
  arma::mat alias;

  {
    arma::mat data(10, 10, arma::fill::randu);
    alias = arma::mat(data.memptr(), 10, 10, false, true);
  }

  // This is invalid!  `data` has been freed and so `alias.memptr()` is no longer valid.
  std::cout << alias;
}
```

Anyway, my best guess is something like that is happening somewhere in the code.  Tracking down exactly where or how is pretty hard given the amounts of indirection in the RNN code though. :smile:

Comment (kartikdutt18) [reply src/mlpack/tests/recurrent_network_test.cpp]: Agreed that does make sense, I will try to get track where variables are assigned logically and also using valgrind. Hopefully I find something that fixes this.

**Comment by rcurtin** [comment]
This code touches the ANN code, and there was recently a big refactoring of the ANN code in #2259, so be sure to merge the master branch into your branch here to make sure that nothing will fail if this PR is merged. :+1: (I'm pasting this message into all possibly relevant PRs.  I think you may have already done that here, but anyway, the message can't hurt. :))

File: src/mlpack/tests/recurrent_network_test.cpp (hunk 4 above)
Comment (kartikdutt18) [reply src/mlpack/tests/recurrent_network_test.cpp]: Hey @rcurtin, sorry for the delayed response, but I am at loss with why the invalid read occurs. I have tried creating a copy anywhere that might lead to an error. I went through how the function calls were made and incase we were using variables that were not accessible, I didn't find any but I am going through them again 😔. Any suggestions on how do I go about investigating this further. I have been running valgrind for the past couple of days with no results to show 😔. I would really appreciate any help or suggestions. Thanks.

File: src/mlpack/tests/recurrent_network_test.cpp (hunk 3 above)
Comment (zoq) [reply src/mlpack/tests/recurrent_network_test.cpp]: I see, thanks for the clarification.

**Pull request closed without merging**